lib/matplotlib/axis.py

```diff
@@ -124,16 +125,54 @@ def __init__(
                 zorder = mlines.Line2D.zorder
         self._zorder = zorder
 
-        grid_color = mpl._val_or_rc(grid_color, "grid.color")
-        grid_linestyle = mpl._val_or_rc(grid_linestyle, "grid.linestyle")
-        grid_linewidth = mpl._val_or_rc(grid_linewidth, "grid.linewidth")
+        grid_color = (
+            mpl._val_or_rc(
+                grid_color, f"grid.{xaxis_yaxis}.{major_minor}.color"
+            )
+            if mpl.rcParams[f"grid.{xaxis_yaxis}.{major_minor}.color"] is not None
+            else (
+                mpl._val_or_rc(grid_color, f"grid.{major_minor}.color")
+                if mpl.rcParams[f"grid.{major_minor}.color"] is not None
+                else mpl._val_or_rc(grid_color, "grid.color")
+            )
```

Comment:
Optional: We have a number of these patterns. You could expand `_val_or_rc` to support a sequence of names (`_val_or_rc(val, *rc_names)`) and resolve to the first not-None, so that one could write this as:
```
grid_color = mpl._val_or_rc(
    grid_color,
    f"grid.{xaxis_yaxis}.{major_minor}.color",
    f"grid.{major_minor}.color"
)